### PR TITLE
feat: Add redis cache storage

### LINKED ISSUE #637: [Feature Request] Integrate Redis as cache storage
### Required prerequisites

- [X] I have searched the [Issue Tracker](https://github.com/camel-ai/camel/issues) and [Discussions](https://github.com/camel-ai/camel/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [ ] Consider asking first in a [Discussion](https://github.com/camel-ai/camel/discussions/new).

### Motivation

we need to record LLM cache into one storage, if needed we can retrieve the content back

### Solution

https://redis.io/

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by koch3092**
## Description

Describe your changes in detail.

## Motivation and Context

Why is this change required? What problem does it solve?
`close #637`

- [x] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/storages/cache_storages/redis.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+
+import json
+import logging
+from typing import Any, Optional
+
+from redis.asyncio import Redis
```
Comment (Wendong-Fan) [inline camel/storages/cache_storages/redis.py:19]:
```suggestion
from typing import TYPE_CHECKING, Any, Optional

if TYPE_CHECKING:
    from redis.asyncio import Redis
```

File: camel/storages/cache_storages/redis.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        Args:
+            url (str): The URL for connecting to the Redis server.
+            **kwargs: Additional keyword arguments for Redis client configuration.
+        """
+        import redis.asyncio as aredis
+
+        self._client = aredis.from_url(url, **kwargs)
+
+    @property
+    def client(self) -> Redis:
```
Comment (Wendong-Fan) [inline camel/storages/cache_storages/redis.py:85]:
```suggestion
    def client(self) -> "Redis":
```

File: camel/storages/cache_storages/redis.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    ) -> None:
+        r"""Creates the Redis client with the provided URL and options.
+
+        Args:
+            url (str): The URL for connecting to the Redis server.
+            **kwargs: Additional keyword arguments for Redis client configuration.
+        """
+        import redis.asyncio as aredis
+
+        self._client = aredis.from_url(url, **kwargs)
```
Comment (Wendong-Fan) [inline camel/storages/cache_storages/redis.py:82]: how about use `self.url` get the value from `__init__`

File: camel/storages/cache_storages/redis.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    def client(self) -> Redis:
+        r"""Returns the Redis client instance.
+
+        Returns:
+            redis.asyncio.Redis: The Redis client instance.
+        """
+        return self._client
+
+    async def set_cache(self, key: str, value: Any, expire: Optional[int] = None) -> bool:
+        r"""Sets a cache entry with an optional expiration time.
```
Comment (Wendong-Fan) [inline camel/storages/cache_storages/redis.py:94]: when set the cache, for key value, how about support adding prefixed value before the key value?

File: pyproject.toml [hunk 5]
```diff
@@ -34,11 +34,12 @@ tiktoken = "^0.7.0"
 colorama = "^0"
 jsonschema = "^4"
 protobuf = "^4"
-pathlib = "^1.0.1"   
+pathlib = "^1.0.1"
 anthropic = "^0.28.0"
 docstring-parser = "^0.15"
 pydantic = ">=1.9,<3"
 curl_cffi = "0.6.2"
+redis = "^5.0.6"
```
Comment (Wendong-Fan) [inline pyproject.toml:42]: let's move redis as optional tool, you can refer to how 'unstructured' set

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @koch3092 ! Overall it's great, I think we can put redis under `camel/storages/key_value_storages`, also it would be great if you can add example code under `examples/storages`, you can refer to `examples/knowledge_graph/neo4j.py`

**Review by zechengz — APPROVED** [review]

**Pull request merged**